### ISSUE TITLE
Difficulty using WithLatestFrom

### ISSUE THREAD

**Issue by jmagaram** [dotnet/reactive#226]
I'm trying to use WithLatestFrom to print words with their respective lengths. My thinking was that if a word got emitted at time X, its length would be calculated at time X+1. I hoped to use WithLatestFrom to pair each calculated length with the most recent word. It doesn't work. I think I know what is happening but am not sure. When a word is emitted, maybe synchronous callbacks notify WithLatestFrom about the length of a word BEFORE it hears about the word. And since the left sequence (length) has a value but the right sequence (word) does not, nothing gets emitted. If I replace the calculated lengths with explicit calls like `word.OnNext("cat"); length.OnNext(3);` it all works fine. 

```
static void Main(string[] args) {
    Subject<string> word = new Subject<string>();
    IObservable<int> length = word.Select(i =>i.Length);
    var report = Observable
        .WithLatestFrom(
            length,
            word,
            (l,w) => new { Word = w, Length = l })
        .Subscribe(Console.WriteLine);
    word.OnNext("cat");
    word.OnNext("donkey");
    word.OnNext("rhinoceros");
    Console.ReadLine();
}
```

The output is:

```
  Cat 6
  Donkey 10
```

**Comment by akarnokd** [comment]
You have to zip them together. As you said, the animal gets onNexted first but at that time, there is no latest from on the other 'rail' and the value gets dropped.

**Comment by danielcweber** [comment]
This is maybe related to #152 and would support my claim that the intuitive behavior of `WithLatestFrom` is to subscribe to the context-observable first, then to the actual source, not vice versa as it is now. But this is just an assumption. Maybe you can merge and test the mentioned pull request locally against your code example ?

**Comment by akarnokd** [comment]
The example uses subjects and the subscription order doesn't matter here. Even if the length is reaches `WithLatestFrom` first, there is no main value so it won't pair up.

**Comment by shiftkey** [comment]
Seconding what @akarnokd suggested, using `Observable.Zip` here seems to show the desired behaviour.

@danielcweber interesting theory: I'll see if I can test that `Observable.WithLatestFrom` example against `master` when I get a chance...

**Comment by jmagaram** [comment]
Zip works in this case but I'm trying to find a general solution where each input generates 0 or 1 output. So if I receive an output, I know it is related to the most recent input.

**Comment by jmagaram** [comment]
I'm trying to get my head around how this Rx stuff works. The order items are emitted seems to be based on the subscription order. It's not intuitive to me. This code...

     word.Subscribe(Console.WriteLine);
     length.Subscribe(Console.WriteLine);
     word.OnNext("cat");
     word.OnNext("donkey");

Generates this result...

     cat
     3
     donkey
     6

But if I swap the subscription order, I get...

     3
     cat
     6
     donkey

**Comment by jmagaram** [comment]
Or for another example, the results change if you swap `number` and `squared` in the `Merge` method.

    Subject<int> number = new Subject<int>();
    IObservable<int> squared = number.Select(i => i * i);
    Observable.Merge(number, squared).Subscribe(Console.WriteLine);
    number.OnNext(4);
    Console.ReadLine();

**Comment by akarnokd** [comment]
Subjects generally use an internal array for the current subscribed `Observers` and they forward signals to them in that specific order.

> word.Subscribe(Console.WriteLine); // obs 1
> length.Subscribe(Console.WriteLine); // obs 2

In this case, you subscribe to the `word` subject first, then subscribe to the `length` `Observable` which in turn subscribes to `word` too. `word` then has `[obs1, obs2]` in its internal list.

> length.Subscribe(Console.WriteLine); // obs 2
> word.Subscribe(Console.WriteLine); // obs 1

If you swap them around, thw length subscribes to `word` and then you subscribe to `word` directly; `word` then has `[obs 2, obs 1]`.

> It's not intuitive to me.

What would be more intuitive to you?

**Comment by jmagaram** [comment]
I'm confused by "you subscribe to `word` first." It looks to me like `length` is the first piece of code to observe `word`.

    Subject<string> word = new Subject<string>();
    IObservable<int> length = word.Select(i => i.Length);
    IDisposable lengthConsole = length.Subscribe(Console.WriteLine);
    IDisposable wordConsole = word.Subscribe(Console.WriteLine);

Reading from top to bottom and filling in those internal arrays...

     word
          has observer 1 length
          has observer 2 wordConsole
     length
          has observer 1 lengthConsole 

So when you emit a word, this happens...

     word
          is sent to observer 1 length
               is sent to observer 1 lengthConsole which prints it
          is sent to observer 2 wordConsole which prints it

So the output would be as follows. Still kind of weird because it looks like the length is being calculated before the word is available.

     3
     cat
     6
     donkey

The output would be exactly the same if the `lengthConsole` and `wordConsole` lines were swapped. This consistency makes sense to me because from a functional perspective (see dependencies below) nothing has changed.

     word = f() // no dependencies
     lengthConsole = f(length)
     wordConsole = f(word)
     length = f(word)

**Comment by akarnokd** [comment]
`Subject`s are lazy when establishing a subscription so even if you assemble sequences in some order, the subscription via `subscribe` determines the actual order in which events in those subjects are dispatched.

> The output would be exactly the same if the lengthConsole and wordConsole lines were swapped.

In pure mathematics perhaps but this is computer science, here there are ordering of events, especially on a single thread, and things attract side-effects pretty quick - such as printing to the console.

**Comment by jmagaram** [comment]
Ok I understand now why I'm not getting the results I want. At the bottom is another version of the original problem that works as expected, but only because I inserted an artificial delay before each result is published.

I'm using Rx for view models. The user manipulates various properties while filling in forms. I do calculations on those properties, like `fullName = firstName + lastName`. Sometimes I do simple synchronous validation, like checking if an email address matches a regular expression pattern. In those situations, Zip can be used to correlate inputs and outputs. But if someone is choosing a new username I might do a web query to validate whether the username is taken. Maybe I'll throttle the input so the check doesn't happen on every keystroke. `Zip` won't work in that case but `WithLatestFrom` or some kind of `Join` might. If I want a GENERAL solution for correlating inputs and outputs, WITHOUT having to know how those calculation occur, I'm kind of stuck. I think my best option is to make sure that all my function outputs INCLUDE their associated inputs, like `new { Email = "justin@google.com", IsFormattedCorrectly = true}` and `new { Url = "www.google.com", IsWebsiteUp = true }`. That way I won't have to correlate the input and output streams. Alternatively I could introduce artificial delays to make every calculation asynchronous use something like `WithLatestFrom` everywhere.

    Subject<string> word = new Subject<string>();
    IObservable<string> wordPub = 
        word
        .Select(i => Observable.Return(i).Delay(TimeSpan.FromSeconds(1)))
        .Concat();
    IObservable<int> lengthPub = 
        wordPub
        .Select(i => Observable.Return(i.Length).Delay(TimeSpan.FromSeconds(0.5)))
        .Concat();
    Observable
        .WithLatestFrom(lengthPub, wordPub, (l, w) => new { Word = w, Length = l })
        .Subscribe(Console.WriteLine);
    word.OnNext("cat");
    word.OnNext("donkey");
    word.OnNext("rhinoceros");
    Console.ReadLine();

**Comment by jmagaram** [comment]
I opened this bug but I don't have a stake in it anymore. I figured out how to work around the behavior. It seems like the current behavior is dependent on the order the method subscribes to the first and second sequences. Because it subscribes to the first sequence first, it learns about length before it hears about the word that produced it. Is this some kind of race condition? From experimentation, I think this is similar to the behavior of Join.

I'm just getting started with rx and will leave it up to the experts to decide if anything needs to change here.
